api/src/main/java/javax/enterprise/inject/spi/builder/ObserverMethodConfigurator.java

```diff
@@ -0,0 +1,197 @@
+     *
+     * @param type of the observed event
+     * @return self
+     */
+    ObserverMethodConfigurator<T> observedType(Type type);
+
+    /**
+     *
+     * Add the qualifier to the observed event
+     * If the builder declares the {@link Default} qualifier, it's automatically removed.
+     *
+     * @param qualifier to add to event
+     * @return self
+     */
+    ObserverMethodConfigurator<T> addQualifier(Annotation qualifier);
+
+    /**
+     *
+     * Add all the qualifiers to the Observed event
+     * If the builder declares the {@link Default} qualifier, it's automatically removed.
+     *
+     * @param qualifiers to add to event
+     * @return self
+     */
+    ObserverMethodConfigurator<T> addQualifiers(Annotation... qualifiers);
+
+    /**
+     *
+     * Add all the qualifiers to the Observed event
+     * If the builder declares the {@link Default} qualifier, it's automatically removed.
+     *
+     * @param qualifiers to add to event
+     * @return self
+     */
+    ObserverMethodConfigurator<T> addQualifiers(Set<Annotation> qualifiers);
+
+    /**
+     * Replace all qualifiers on the Observed event.
+     * If the builder declares the {@link Default} qualifier, it's automatically removed.
+     *
+     * @param qualifiers to put on event
+     * @return self
+     */
+    ObserverMethodConfigurator<T> qualifiers(Annotation... qualifiers);
+
+    /**
+     * Replace all qualifiers on the Observed event.
+     * If the builder declares the {@link Default} qualifier, it's automatically removed.
+     *
+     * @param qualifiers to put on event
+     * @return self
+     */
+    ObserverMethodConfigurator<T> qualifiers(Set<Annotation> qualifiers);
+
+    /**
+     *
+     * Set the {@link Reception} mode for the observer to build
+     *
+     * @param reception reception type
+     * @return self
+     */
+    ObserverMethodConfigurator<T> reception(Reception reception);
+
+    /**
+     *
+     * Set the {@link TransactionPhase} for the observer to build
+     *
+     * @param transactionPhase phase for the observer
+     * @return self
+     */
+    ObserverMethodConfigurator<T> transactionPhase(TransactionPhase transactionPhase);
+
+    /**
+     *
+     * Set the priority for the observer to build
+     *
+     * @param priority priority of the observer
+     * @return self
+     */
+    ObserverMethodConfigurator<T> priority(int priority);
+
+   /**
+    * Define consumer to call when event is notified
+    *
+    * @param consumer to call for the event notification
+    * @return self
+    */
+   ObserverMethodConfigurator<T> notifyWith(Consumer<T> consumer);
+
+   /**
+    *
+    * Define notification directly in the observer with event metadata and an Instance for beans
+    *
+    * @param metadata metadata of the event
+    * @param instance instance to retrieve all beans
+    * @return self
+     */
+   ObserverMethodConfigurator<T> notifyWith(EventMetadata metadata, Instance<Object> instance);
```

Comment:
Eh, not `BiFunction` but a consumer with three arguments. But it doesn't exist in JDK. Well, theoretically we could spit this in two methods `notifyWith(BiConsumer<EventMetadata, T> consumer)` and `notifyWith(BiConsumer<Instance<Object>, T> consumer)` (here the Instance could be used to obtain `EventMetadata`, although I'm not sure it would work).